replicator/main.go

```diff
@@ -0,0 +1,847 @@
+)
+
+// TODO: Escape field names?
+// TODO: Tuple support
+
+var showTimestamps = true
+var debugQueries = false
+var retryCount = 20
+var maxWaitBetweenRetries = 5 * time.Second
+
+func main() {
+	var (
+		keyspace    string
+		table       string
+		source      string
+		destination string
+		consistency string
+	)
+
+	flag.StringVar(&keyspace, "keyspace", "", "keyspace name")
+	flag.StringVar(&table, "table", "", "table name")
+	flag.StringVar(&source, "source", "", "address of a node in source cluster")
+	flag.StringVar(&destination, "destination", "", "address of a node in destination cluster")
+	flag.StringVar(&consistency, "consistency", "", "consistency level (one, quorum, all)")
+	flag.String("mode", "", "mode (ignored)")
+	flag.Parse()
+
+	cl := gocql.One
+	switch strings.ToLower(consistency) {
+	case "one":
+		cl = gocql.One
+	case "quorum":
+		cl = gocql.Quorum
+	case "all":
+		cl = gocql.All
+	}
+
+	adv := scylla_cdc.AdvancedReaderConfig{
+		ConfidenceWindowSize:   30 * time.Second,
+		ChangeAgeLimit:         24 * time.Hour,
+		QueryTimeWindowSize:    24 * time.Hour,
+		PostEmptyQueryDelay:    30 * time.Second,
+		PostNonEmptyQueryDelay: 10 * time.Second,
+		PostFailedQueryDelay:   1 * time.Second,
+	}
+
+	reader, rowsRead, err := MakeReplicator(
+		source, destination,
+		[]string{keyspace + "." + table},
+		&adv,
+		cl,
+	)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// React to Ctrl+C signal, and stop gracefully after the first signal
+	// Second signal cancels the context, so that the replicator
+	// should stop immediately, but still gracefully
+	// The third signal kills the process
+	signalC := make(chan os.Signal)
+	go func() {
+		<-signalC
+		now := time.Now()
+		log.Printf("stopping at %v", now)
+		reader.StopAt(now)
+
+		<-signalC
+		log.Printf("stopping now")
+		cancel()
+
+		<-signalC
+		log.Printf("killing")
+		os.Exit(1)
+	}()
+	signal.Notify(signalC, os.Interrupt)
+
+	if err := reader.Run(ctx); err != nil {
+		log.Println(err)
+		os.Exit(1)
+	}
+
+	log.Printf("quitting, rows read: %d", *rowsRead)
+}
+
+func MakeReplicator(
+	source, destination string,
+	tableNames []string,
+	advancedParams *scylla_cdc.AdvancedReaderConfig,
+	consistency gocql.Consistency,
+) (*scylla_cdc.Reader, *int64, error) {
+	// Configure a session for the destination cluster
+	destinationCluster := gocql.NewCluster(destination)
+	destinationSession, err := destinationCluster.CreateSession()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tracker := scylla_cdc.NewClusterStateTracker(gocql.TokenAwareHostPolicy(gocql.RoundRobinHostPolicy()))
+
+	// Configure a session
+	cluster := gocql.NewCluster(source)
+	cluster.PoolConfig.HostSelectionPolicy = tracker
+	session, err := cluster.CreateSession()
+	if err != nil {
+		destinationSession.Close()
+		return nil, nil, err
+	}
+
+	progressManager, err := scylla_cdc.NewTableBackedProgressManager(session, "ks.cdc_go_replicator_progress")
```

Comment:
hm, if we run multiple replicators in parallel, they will break each others' progress managers
this could be solved by assigning some ID to a replicator instance; the table name would include this ID
restarting an instance would require passing the ID